ethcore/src/block.rs

```diff
@@ -168,16 +168,17 @@ impl<'x> OpenBlock<'x> {
 	/// Push a transaction into the block.
 	///
 	/// If valid, it will be executed, and archived together with the receipt.
-	pub fn push_transaction(&mut self, t: SignedTransaction) -> Result<&Receipt, Error> {
+	pub fn push_transaction(&mut self, t: &SignedTransaction) -> Result<&Receipt, Error> {
```

Comment:
The transaction is still cloned if the else branch is taken, are you doing this because of saving an allocation if the if branch is taken?

Not sure how this plays along w.r.t. to `speculative execution`  on modern CPUs

Reply:
Wait, which `else`-branch do you mean?

Reply:
haha sorry I meant the implicit else branch, i.e, if https://github.com/OpenEthereum/open-ethereum/pull/11531/files#diff-e2e171033e005c3b329818aaff213767R172-R174 is missed

Reply:
I see, well there's that also, but mainly passing the tx by ref is fallout from a borrow from higher up in the call chain, in `enact_verified()`. (I was also unsure wth the `into()` was doing).

Reply:
The `into()` was not doing anything

Reply:
I'm also not sure why we need this change, if anything, it makes more allocations.

Reply:
So in [`import_verified_blocks()`](https://github.com/OpenEthereum/open-ethereum/blob/7d54e9258dd04671e6204769e1f624b7dbf496d0/ethcore/src/client/client.rs#L279) we call [`check_and_lock_block()`](https://github.com/OpenEthereum/open-ethereum/blob/7d54e9258dd04671e6204769e1f624b7dbf496d0/ethcore/src/client/client.rs#L365) passing in an owned `PreverifiedBlock`, so to make the rest of the code work there we do [some cloning](https://github.com/OpenEthereum/open-ethereum/blob/7d54e9258dd04671e6204769e1f624b7dbf496d0/ethcore/src/client/client.rs#L301-L302) that didn't seem necessary and that's where this PR started. 
When we call [`enact_verified()`](https://github.com/OpenEthereum/open-ethereum/blob/62b73a64604c7326a8fc6cb9b553336903aaeca1/ethcore/src/block.rs#L462) we move the block but what is actually needed for `enact()` is actually the [header, transactions and the (max) 2 uncles](https://github.com/OpenEthereum/open-ethereum/blob/62b73a64604c7326a8fc6cb9b553336903aaeca1/ethcore/src/block.rs#L474-L476).
What I'd __really__ would like to do is have a way to split up the `PreverifiedBlock` into (owned) components and be able to move them independently. I don't know of a way to do that, and that's why I ended up with this: it avoids cloning the block `bytes` in `import_verified_blocks()` and a few `Header` structs, and passes references around instead. Maybe this mean more clones, as each transaction that is added to the block is cloned, but they are also smaller (I hope).

Reply:
I guess you could use some `mem::take()` tricks, for example, replace Bytes with an empty Vec.

But, why can't you take `PreverifiedBlock` by value instead in `enact`? As I did here: https://github.com/OpenEthereum/open-ethereum/compare/na-ethcore-blockerror#diff-e2e171033e005c3b329818aaff213767R411-R564

I'm not saying that it is elegant but it worked for me at least :)

Reply:
This sounds thing reminds me of something :P 
> The real problem is that programmers have spent far too much time worrying about efficiency in the wrong places and at the wrong times; premature optimization is the root of all evil (or at least most of it) in programming.”

Seriously though, 
> What I'd really would like to do is have a way to split up the PreverifiedBlock into (owned) components and be able to move them independently. I don't know of a way to do that

create a type for parts or pass them directly instead of `PreferifiedBlock`? or what Niklas suggested

Reply:
Ok, thank you for the input, much appreciated. At the end I went with `mem::take()`-ing the block RLP bytes and after some further shuffling things around I think I have something decent.